packages/rest/src/router/metadata.ts

```diff
@@ -47,95 +50,81 @@ export function api(spec: ControllerSpec) {
       typeof constructor === 'function',
       'The @api decorator can be applied to constructors only.',
     );
-    Reflector.defineMetadata('loopback:api-spec', spec, constructor);
+    const apiSepc = resolveControllerSpec(constructor, spec);
+    Reflector.defineMetadata(API_SPEC_KEY, apiSepc, constructor);
   };
 }
 
 interface RestEndpoint {
   verb: string;
   path: string;
+  spec?: OperationObject;
+  target: any;
 }
 
-export function getControllerSpec(constructor: Function): ControllerSpec {
+function resolveControllerSpec(
+  constructor: Function,
+  spec?: ControllerSpec,
+): ControllerSpec {
   debug(`Retrieving OpenAPI specification for controller ${constructor.name}`);
 
-  let spec: ControllerSpec = Reflector.getMetadata(
-    'loopback:api-spec',
-    constructor,
-  );
-
   if (spec) {
     debug('  using class-level spec defined via @api()', spec);
-    return spec;
-  }
-
-  spec = {paths: {}};
-  for (
-    let proto = constructor.prototype;
-    proto && proto !== Object.prototype;
-    proto = Object.getPrototypeOf(proto)
-  ) {
-    addPrototypeMethodsToSpec(spec, proto);
+    spec = Object.assign({}, spec);
+  } else {
+    spec = {paths: {}};
   }
-  return spec;
-}
 
-function addPrototypeMethodsToSpec(spec: ControllerSpec, proto: any) {
-  const controllerMethods = Object.getOwnPropertyNames(proto).filter(
-    key => key !== 'constructor' && typeof proto[key] === 'function',
-  );
-  for (const methodName of controllerMethods) {
-    addControllerMethodToSpec(spec, proto, methodName);
-  }
-}
+  const endpoints =
+    Reflector.getMetadata(ENDPOINTS_KEY, constructor.prototype) || {};
+
+  for (const op in endpoints) {
+    const endpoint = endpoints[op];
+    const className =
+      endpoint.target.constructor.name ||
```

Comment:
Is there a case where `endpoint` could exist but not have a `target` property? (Looking to avoid seeing `Cannot read property "constructor" of undefined`)

Reply:
There is no chance as `target` will be always available from the decorator applications.